src/client/Microsoft.Identity.Client/OAuth2/Throttling/UiRequiredProvider.cs

```diff
@@ -0,0 +1,118 @@
+                var entry = new ThrottlingCacheEntry(ex, s_uiRequiredExpiration);
+                ThrottlingCache.AddAndCleanup(thumbprint, entry, logger);
+            }
+        }
+
+        public void ResetCache()
+        {
+            ThrottlingCache.Clear();
+        }
+
+        public void TryThrottle(AuthenticationRequestParameters requestParams, IReadOnlyDictionary<string, string> bodyParams)
+        {
+            if (!ThrottlingCache.IsEmpty() && IsSupported(requestParams))
+            {
+                var logger = requestParams.RequestContext.Logger;
+
+                string fullThumbprint = GetRequestStrictThumbprint(
+                    bodyParams,
+                    requestParams.AuthorityInfo.CanonicalAuthority,
+                    requestParams.RequestContext.ServiceBundle.PlatformProxy.CryptographyManager);
+
+                TryThrowException(fullThumbprint, logger);
+            }
+        }
+
+        private void TryThrowException(string thumbprint, ICoreLogger logger)
+        {
+            if (ThrottlingCache.TryGetOrRemoveExpired(thumbprint, logger, out MsalServiceException ex) &&
+                ex is MsalUiRequiredException uiException)
+            {
+                logger.WarningPii(
+                    $"[Throttling] Exception thrown because of throttling rule UiRequired - thumbprint: {thumbprint}",
+                    $"[Throttling] Exception thrown because of throttling rule UiRequired ");
+
+                // mark the exception for logging purposes                
+                throw new MsalThrottledUiRequiredException(uiException);
+            }
+        }
+
+        private static bool IsSupported(AuthenticationRequestParameters requestParams)
+        {            
+            return !requestParams.IsConfidentialClient &&
+                requestParams.ApiId == TelemetryCore.Internal.Events.ApiEvent.ApiIds.AcquireTokenSilent;
+        }
+
+        /// <summary>
+        /// The strict thumbprint is based on: 
+        /// ClientId
+        /// Authority (env + tenant)
+        /// Scopes
+        /// hash(RT) or UPN for IWA (not supported)
+        /// </summary>
+        private static string GetRequestStrictThumbprint(
+            IReadOnlyDictionary<string, string> bodyParams,
+            string authority,
+            ICryptographyManager crypto)
+        {
+            StringBuilder sb = new StringBuilder();
+            if (bodyParams.TryGetValue(OAuth2Parameter.ClientId, out string clientId))
+            {
+                sb.Append((clientId ?? "") + ThrottleCommon.KeyDelimiter);
+            }
+            sb.Append(authority + ThrottleCommon.KeyDelimiter);
+            if (bodyParams.TryGetValue(OAuth2Parameter.Scope, out string scopes))
+            {
+                sb.Append((scopes ?? "") + ThrottleCommon.KeyDelimiter);
+            }
+
+            if (bodyParams.TryGetValue(OAuth2Parameter.RefreshToken, out string rt) && 
```

Comment:
so it looks like we are using the RT to create a thumbprint for each of the cached exceptions. How does this work with the android broker because it does not return an RT?

Reply:
This is a great question! I was initially thinking to check if "IsBrokerRequest" but then I realized that broker requests do not use the "TokenClient". So we should be good. 

The broker requests communicate with brokers through different methods, not using HTTP calls. TokenClient class, where all the throttling methods are, are not used by broker.